presto-main/src/main/java/io/prestosql/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -153,16 +151,16 @@ private boolean exploreNode(int group, Context context)
 
     private <T> Rule.Result transform(PlanNode node, Rule<T> rule, Context context)
     {
-        Capture<T> nodeCapture = newCapture();
-        Pattern<T> pattern = rule.getPattern().capturedAs(nodeCapture);
+        Pattern<T> pattern = rule.getPattern();
         Iterator<Match> matches = pattern.match(node, context.lookup).iterator();
         while (matches.hasNext()) {
             Match match = matches.next();
             long duration;
             Rule.Result result;
             try {
                 long start = System.nanoTime();
-                result = rule.apply(match.capture(nodeCapture), match.captures(), ruleContext(context));
+                //noinspection unchecked
+                result = rule.apply((T) node, match.captures(), ruleContext(context));
```

Comment:
It wasn't fully redundant. 

I would prefer to leave it as it was. If we are going to remove static typing from pattern matching then we should do it entirely. Personally I have mixed feelings. I see a benefit, but also I see a loss. I am not sure what is better.

Reply:
The capture is not adding any additional static typing guarantees.
Instead, it obfuscated the fact that we invoke the rule on the `node` -- instead of passing the node explicitly, we pull it from somewhere else.

Reply:
> The capture is not adding any additional static typing guarantees.

IMO, it does, it works like "`instenceof`" where pattern contains the class information and pattern.

Reply:
You didn't convince me it's worth it (the indirection is not nice). But you are right.